ui/app/components/auth/form/oidc-jwt.ts

```diff
@@ -62,188 +58,193 @@ export default class AuthFormOidcJwt extends AuthBase {
+      );
+      this.authUrl = data.authUrl;
       this.isOIDC = true;
     } catch (e) {
-      const { httpStatus } = e as AdapterError;
-      const message = errorMessage(e);
-      // track errors but they only display on submit
+      const { status, message } = await this.api.parseError(e);
+      // errors are tracked here but they only display on submit
       this.errorMessage =
-        httpStatus === 400 ? 'Invalid role. Please try again.' : `Error fetching role: ${message}`;
-      // if the mount is configured for JWT authentication via static keys, JWKS, or OIDC discovery
+        // A 400 is returned if OIDC is configured but does not have a default role set.
+        status === 400 ? 'Invalid role. Please try again.' : `Error fetching role: ${message}`;
+      // If the mount is configured for JWT authentication via static keys, JWKS, or OIDC discovery
       // this specific error is returned. Flip the isOIDC boolean accordingly, otherwise assume OIDC.
-      this.isOIDC = message !== ERROR_JWT_LOGIN;
+      this.isOIDC = !message.includes(ERROR_JWT_LOGIN);
     }
   });
 
-  login = task(
-    waitFor(async (submitData) => {
-      if (this.isOIDC) {
-        this.startOIDCAuth();
-      } else {
-        this.continueLogin(submitData);
-      }
-    })
-  );
+  generateRedirectUri(namespace = '', path = '') {
+    const origin = window.location.origin;
+    const qp = namespace ? { namespace } : {};
+    const routeUrl = this.router.urlFor(
+      'vault.cluster.oidc-callback',
+      { auth_path: path },
+      { queryParams: qp }
+    );
+    return `${origin}${routeUrl}`;
+  }
 
-  async continueLogin(data: JwtLoginData | OidcLoginData) {
-    try {
-      // TODO CMB backend should probably be path, but holding off refactor since api service may remove need all together
-      // OIDC callback returns a token so authenticate with that
-      const backend = this.isOIDC && 'token' in data ? 'token' : this.args.authType;
-
-      const authResponse = await this.auth.authenticate({
-        clusterId: this.args.cluster.id,
-        backend,
-        data,
-        selectedAuth: this.args.authType,
-      });
-
-      // responsible for redirect after auth data is persisted
-      this.handleAuthResponse(authResponse);
-    } catch (error) {
-      this.onError(error as Error);
+  // * LOGIN WORKFLOW BEGINS
+  async loginRequest(formData: JwtLoginData) {
+    if (this.isOIDC) {
+      return await this.loginOidc();
+    } else {
+      return await this.loginJwt(formData);
+    }
+  }
+
+  async loginJwt(formData: JwtLoginData) {
+    const { path, jwt, role } = formData;
+    const { auth } = <JwtOidcLoginApiResponse>await this.api.auth.jwtLogin(path, { jwt, role });
+    // displayName is not returned by auth response and is set in persistAuthData
+    return this.normalizeAuthResponse(auth, { path, tokenKey: 'clientToken', ttlKey: 'leaseDuration' });
+  }
+
+  async loginOidc() {
```

Comment:
I refactored the logic in this component a bit so that all of the OIDC methods return something and are called here. Hopefully this makes the logic a little easier to parse. This also has the benefit of consolidating the error handling so most failures bubble up to the parent naturally, aside from the cases that need special handling (like closing the popup window)